modules/http/http-loadbalancer.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2018 Balazs Scheidler
+ * Copyright (c) 2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef HTTP_LOADBALANCER_H_INCLUDED
+#define HTTP_LOADBALANCER_H_INCLUDED 1
+
+#include "syslog-ng.h"
+
+typedef enum
+{
+  HTTP_TARGET_OPERATIONAL,
+  HTTP_TARGET_FAILED
+} HTTPLoadBalancerTargetState;
+
+typedef struct _HTTPLoadBalancerTarget HTTPLoadBalancerTarget;
+typedef struct _HTTPLoadBalancerClient HTTPLoadBalancerClient;
+typedef struct _HTTPLoadBalancer HTTPLoadBalancer;
+
+
+/* NOTE: this struct represents an actual HTTP target URL.  The existance of
+ * this structure is ensured even in multi-threaded environments.  Some of
+ * the members of the struct are read-only and are _always_ available
+ * without locking.  Others are protected by the LoadBalancer's lock and
+ * should be manipulated under the protection of that lock.
+ */
+struct _HTTPLoadBalancerTarget
+{
+  /* read-only data, no need to lock */
+  gchar *url;
+  gint index;
+  /* read-write data, locking must be in effect */
+  HTTPLoadBalancerTargetState state;
+  gint number_of_clients;
+  gint max_clients;
+  time_t last_failure_time;
+};
+
+struct _HTTPLoadBalancerClient
+{
+  HTTPLoadBalancerTarget *target;
+};
+
+void http_lb_client_init(HTTPLoadBalancerClient *, HTTPLoadBalancer *lb);
+void http_lb_client_deinit(HTTPLoadBalancerClient *);
+
+struct _HTTPLoadBalancer
+{
+  GStaticMutex lock;
+  HTTPLoadBalancerTarget *targets;
+  gint num_targets;
+  gint num_clients;
```

Comment:
just a note (to other reviewers): 
`num_clients` is the number of workers (when I read the code, I always had to remember myself to this fact, otherwise it is easier to lost in num_* variables ). I understand that you don't want to connect the LoadBalancer class to the workers, but the num_clients maybe too 'sterile'

Reply:
it's the number of HTTPLoadBalancerClient instances, in this way it is really the number of clients. Whether we have threads in this case is not a requirement.

I'd leave this as is.